content/docs/instrumenting/http_sd.md

```diff
@@ -0,0 +1,90 @@
+---
+title: HTTP SD
+sort_rank: 7
+---
+
+# Writing HTTP Service Discovery
+
+Prometheus provides a generic [HTTP Service Discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config),
+that enables to discover targets over an HTTP endpoint.
+
+The HTTP Service Discovery is complimentary to the supported service
+discoveries, and is an alternative to the [File-based Service Discovery](https://prometheus.io/docs/guides/file-sd/#use-file-based-service-discovery-to-discover-scrape-targets).
+
+## Comparison between File-Based SD and HTTP SD
+
+Here is a table comparing our two generic Service Discovery implementations.
+
+| Item | File SD | HTTP SD |
+| ---- | ------- | ------- |
+| Event Based | Yes, via inotify | No |
+| Update frequency | Instant, thanks to inotify | Following refresh_interval |
+| Format | Yaml or JSON | JSON |
+| Transport | Local file | HTTP/HTTPS |
+| Security | File-Based security | TLS, Basic auth, Authorization header, OAuth2 |
+
+## Requirements of HTTP SD endpoints
+
+The HTTP SD endpoints should have been written specifically for Prometheus.
+The response is consumed as is, unmodified. On each refresh interval (default: 1
+minute), Prometheus will do a GET request to the HTTP SD endpoint. The GET
+request contains a `X-Prometheus-Refresh-Interval-Seconds` HTTP header with the
+refresh interval.
+
+The SD endpoint must answer with an HTTP 200 response, with the HTTP Header
+`Content-Type: application/json`. The answer must be UTF-8 formatted.
+In the case when no targets are returned, an HTTP 200 must also be emitted, with
+an empty list `[]`.
+
+It is expected that the whole list of targets is returned on every scrape. There
+is no client-side caching mechanism. A prometheus instance does not send its
```

Comment:
Does this imply that a failure (4xx, 5xx, TCP RST) will empty any existing discovery data? If any failure results in no targets, we should probably call this out.

Reply:
It should keep the previous targets set. I will update.

Reply:
Rather than "no client-side cache", I would say "It does not support incremental updates".

Reply:
Also "It does not store data between restarts"